lib/puppet/resource_api/property.rb

```diff
@@ -69,10 +86,40 @@ def rs_value
   # method overloaded only for the :ensure property, add option to check if the
   # rs_value matches is. Only if the class is child of
   # Puppet::ResourceApi::Property.
-  def def_insync?
+  def def_ensure_insync?
     define_singleton_method(:insync?) { |is| rs_value.to_s == is.to_s }
   end
 
+  def def_custom_insync?
+    define_singleton_method(:insync?) do |is|
+      provider    = @referrable_type.my_provider
+      context     = @referrable_type.context
+      should_hash = @resource.rsapi_canonicalized_target_state
+      is_hash     = @resource.rsapi_current_state
+      title       = @resource.rsapi_title
+
+      raise(Puppet::DevError, 'No insync? method defined in the provider; an insync? method must be defined if the custom_insync feature is defined for the type') unless provider.respond_to?(:insync?)
+      provider_insync_result = provider.insync?(context, title, @attribute_name, is_hash, should_hash)
+      # if nil, super and rely on Puppet::Property.insync?
+      # if true/false, return that and use the default change messaging
+      # if string, set value of custom change reporting and treat as false
```

Comment:
It's surprising to me that the `insync?` predicate method can return so many different values, and due to the lack of types in ruby, a provider could accidentally return string `"true"` instead of boolean `true`, and be treated as false. Similar confusion can result if the provider returns symbol `:true`:

```
irb(main):006:0> :true == true
=> false
```


It might be better to return two values instead:

```ruby
result, message = provider.insync(...)
```

where `result` is always `true/false` and `message` is always a non-nil string when `result` is false.

```ruby
irb(main):001:1* def insync?
irb(main):002:1*   true
irb(main):003:0> end
=> :insync?
irb(main):004:0> result, message = insync?
=> true
irb(main):005:0> message
=> nil
```

Ideally I think it would be better to return an `InsyncResult` object, but that might be a breaking API change for existing RSAPI providers?

Reply:
For this issue, I'm fine if `insync?` returns tri-state true/false/nil. But it seems strange to force the implementation to return changed message when the resource is insync. See my comment about `change_to_s` below.

Reply:
I agree that the current return value handling should be more strict. As it stands values that are not `true` and not `is_a?(String)` could slip through.

I'm undecided whether anything other than literal `true` should cause a change notification exposing the return value or whether return values not in the set of `true`, `false`, `nil` and `is_a?(String)` should cause an error like https://github.com/puppetlabs/puppet-resource_api/blob/dab8aa965899af8b30095445dab73f850779f8d2/lib/puppet/resource_api.rb#L354-L361. I'm leaning more towards the exposing approach as it is easy to catch in visual inspection of puppet output (what people commonly do to test their code), and we set `strict` to the wrong level in the default templates for unit testing (https://github.com/puppetlabs/pdk-templates/blob/377df33b66ce9db9bdd55a1b6b7d3bf5ec887b39/config_defaults.yml#L634)

Independent of that, I also like the idea of the two-value return approach, specifically because it allows the additional possibility of returning `[nil, 'change message']` for values that use the default comparison, but require a custom change message. Even in this case, a more strict approach for checking the return types should be used.

PS: Specifically to the examples of `:true` and `"true"`: I'd dearly love to enable `Lint/BooleanSymbol` rubocop, which traditionally has been turned off because of the usage of `:true` in low-level APIs. In Resource API providers, folks never should need to use `:true` or `:false`.

Reply:
Just to reiterate, I'd really like to avoid the magic that "boolean true means success, but string means failure and is the message". I think it's better to be explicit about what is control (true/false/nil) vs data (message).